### ISSUE TITLE
FSDP: automatic instantiation on GPUs (aka zero.Init)

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#120878]
### 🚀 The feature, motivation and pitch

moving from [slack discussion](https://pytorch.slack.com/archives/C3PDTEV8E/p1709153741504309) 

How can FSDP shard a model directly to GPUs during model instantiation?

This is critical for instantiating huge models.

### Alternatives

In Deepspeed this exists since day 0 of the framework using [`zero.Init`](https://deepspeed.readthedocs.io/en/latest/zero3.html#constructing-massive-models)

This feature is integrated into HF Transformers' `from_pretrained` that instantiates the model directly on GPUs - [here](https://github.com/huggingface/transformers/blob/2209b7afa04b3a6366350065f541e9248d6663c2/src/transformers/modeling_utils.py#L1301)

There is an emerging need to control when this should be activated and when
not as we now have many situations with multiple concurrent models and this
current setup no longer works well.So I thought that HF Transformers should
gather at least some needs of various frameworks - Deepspeed ZeRO, FSDP and
may be others where `from_pretrained` provides hooks into the model
instantiation stage. The 2 needs are:  

  1. yes/no activate flag
  2. framework provided context to instantiate in - so `zero.Init` would be just one of them

The intention is that the user won't need to do anything special when they
create the model. The framework would register with HF Transformers everything
it needs to do it smoothly and correctly.

### Additional context

on slack Andrew Gu posted:

> If you specify the  `device_id` arg, FSDP should move the unsharded module to GPU before sharding so that the sharding ops are fast (but still requiring CPU init and copy H2D). Otherwise, the only other way to initialize _and_ shard on GPU is to do so-called “manual wrapping”, which was the common approach from Fairscale. I do not think that would work with HF transformers.

> For manual wrapping, you have to call `FullyShardedDataParallel(module, …)` on
`module` immediately after it was constructed to avoid spiking memory. This is
intrusive to the model definition and not acceptable in many cases. I am just
mentioning it since it can be the fastest init supported.  

    
    
    # Example manual wrapping with `enable_wrap`/`wrap`
    # Wrap each transformer block and the root
    from torch.distributed.fsdp import enable_wrap, wrap
    
    class Transformer(nn.Module):
      def __init__(self, ...):
        layers = []
        for layer_id in range(num_layers):
          layer = TransformerBlock(...)
          layer = wrap(layer, ...)
          layers.append(layer)
        self.layers = nn.ModuleList(layers)
        
    with torch.device("cuda"):
      with enable_wrap(wrapper_cls=FullyShardedDataParallel, **fsdp_kwargs):
        model = Transformer(...)
        wrap(model)

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @kwen2501

**Comment by rohan-varma** [comment]
@stas00 Have you tried FSDP's (admittedly limit) support for meta device to instantiate the model directly on GPU? For finetuning use cases, you can do something like the following:

```
with torch.device("meta"):
    model = model_call(....)

# wrap in FSDP, avoiding CPU all together and bringing sharded parameters to GPU.
# note that this will fill all parameters with empty / garbage data.
model = FSDP(model, device_id=torch.cuda.current_device(), param_init_fn=lambda m: m.to_empty(recurse=False))

# load state_dict for finetuning use cases
model.load_state_dict(...)
# Note: user is still responsible for initializing components of model that aren't overwritten by state_dict, such as rope embeddings for llama or specific PEFT adapter components.
```

> The intention is that the user won't need to do anything special when they
create the model

Is it possible to offer some sort of interface / wrapper in upstream libraries to hide this complexity from the user?

**Comment by carmocca** [comment]
@rohan-varma `load_state_dict` cannot be used like this for the sharded model, it will require loading a distributed checkpoint (DCP), converting a `torch.save` checkpoint, or summoning the full params before loading (on CPU - provided you have enough RAM).

> Is it possible to offer some sort of interface / wrapper in upstream libraries to hide this complexity from the user?

Since you asked, Lightning Fabric and PyTorch Lightning (disclaimer: I work on this) do this already: https://lightning.ai/docs/fabric/stable/advanced/model_init.html#model-parallel-training-fsdp-and-deepspeed. Including support for loading full and sharded checkpoints.

**Comment by stas00** [comment]
@rohan-varma, yes, this is why I'm trying to discuss this. Since HF Transformers already does a special use-case for Deepspeed, so its sharding to GPUs is automatic for users, but as we now have other similar use cases - I think [that hardcoded solution](https://github.com/huggingface/transformers/blob/2209b7afa04b3a6366350065f541e9248d6663c2/src/transformers/modeling_utils.py#L1301) needs to be upgraded to be supportive of various use-cases. So I think FSDP's need integration would be a perfect use-case. I have tagged Sourab in the comment above who took over the initial work I did at HF Transformers 3 years ago.

**Comment by awgu** [comment]
Similar to https://github.com/pytorch/pytorch/issues/120879#issuecomment-1977140356, I wanted to share what I was thinking as an option for meta-device init in the new FSDP.

The PR is https://github.com/pytorch/pytorch/pull/120351. The idea is that, by representing FSDP sharded parameters as `DTensor`s, we can rely on `DTensor`'s op dispatch to compute the correct randomness with respect to the global (unsharded) shape.

As such, we can shard a meta-device model with TP and FSDP, leaving it on meta-device; materialize the sharded model on GPU (e.g. using `to_empty(device)`); and then either load from a checkpoint or randomly initialize the sharded meta-device parameters through `DTensor` dispatch.

For example, taken from that PR:
```
# Pre-training flow (no checkpoint)
global_mesh = init_device_mesh(..., mesh_dim_names=("dp", "tp"))
dp_mesh, tp_mesh = global_mesh["dp"], global_mesh["tp"]
with torch.device("meta"):
  model = ...
  parallelize_module(model, tp_mesh, ...)
  fully_shard(model, mesh=dp_mesh, ...)
for param in model.parameters():
  assert param.device.type == "meta"

model.to_empty(device="cuda")
random.manual_seed(42, global_mesh)
for module in model.modules():
  if hasattr(module, "reset_parameters"):
    module.reset_parameters()
```

**Comment by mori360** [comment]
Solved by [initial meta-device](https://github.com/pytorch/pytorch/pull/120351)
